require(shiny)
require(DT)
options(shiny.maxRequestSize=200*1024^2)
require(shinyBS)
require(shinyjs)

# Define UI for application
shinyUI(fluidPage(
  tags$style(type="text/css",
             ".shiny-output-error { visibility: hidden; }",
             ".shiny-output-error:before { visibility: hidden; }"),
  withMathJax(),
  useShinyjs(),
  titlePanel(h1("FMT Efficacy Analysis Toolkit v2.1", align = "left"), windowTitle = "FMT Efficacy Analysis Toolkit (FEAT)"),

  navlistPanel(
    tabPanel("Home",
             h3("Description"),
             p("This tool is designed to help quickly and interactively analyze microbiome sequencing data from FMT experiments. The two main inputs that you need are an OTU table and a mapping file."),
             h3("Inputs"),
             p(strong("OTU Table:"), "This tool will work best with manageably-sized OTU tables in the hdf5 biom format. The preferred method of getting a table in this format is to use the sequence data output from split_libraries.py (from QIIME), and dereplicate sequences to generate OTUs. Then, filter out singleton OTUs to drastically reduce the size of the OTU table. For an example on how to do this, and retain sequence information, see ", a("here", href="https://gist.github.com/gregcaporaso/f3c042e5eb806349fa18", target="_blank")),
             p(strong("Mapping File:"), "A mapping file that contains at least one metadata category in which the donor, pre-transplant recipient, and post-transplant recipient can be identified. It is important that the SampleID field (from your standard QIIME mapping file) remains unchanged, as it is the field used to add metadata to the OTU table."),
             h3("Overview"),
             p("This tool will split your OTU table into your experiment-specific samples. Next, a filtering step occurs after the OTU counts are normalized, that will remove OTUs that are less abundant than the given threshold. Subsequently, the tool will identify OTUs that are unique and specific to the donor and to the recipient samples, and perform analyses on the post-transplant samples based only on these information-rich OTUs. This includes calculating a distance matrix of the samples based on these OTUs, and allowing for the interactive visualization of this information plus any additional metadata."),
             p("Moreover, we have developed metrics to help quantify the ability of a microbiota to engraft in a recipient and for a microbiota to be pushed out by another (see figure)."),
             img(src='FEAT-Metrics.png'),
             p("Click through the sidebar to add inputs and change the parameters for analyzing the data.")),

    "Input & Data Pre-Processing",

    tabPanel("Select Inputs & Settings",
             h2("Select Inputs & Settings"),
             fluidRow(column(6,
                             wellPanel(h3("1) Load Mapping File, OTU Table, and Taxonomic Information"),
                                       fileInput("input_otu_table", label = h5("Input OTU Table"), accept = 'Document'),
									   bsPopover('input_otu_table', "OTU Table", "Absolute-filtered OTU table in hdf5 BIOM (.biom) format."),
                                       hr(),
									   fileInput("mapping_file", label = h5("Mapping File"), accept = c('txt', 'text/plain')),
                                       bsPopover("mapping_file", "Mapping File", "This should be the same mapping file used to create your OTU table, and should contain at least one metadata category to select the details of the transplant for your experiment of interest."),
                                       hr(),
                                       span(strong(textOutput('zeroeth_check')), style = "color:green"),
                                       hr(),
                                       h5("Starting Number of OTUs:"),
                                       textOutput("raw_otu_count")									   
                             ),
                             conditionalPanel(condition = "output.zeroeth_check == 'Table loaded successfully...'",
                                              wellPanel(h3("2) Select FMT Samples"),
                                                        uiOutput("compare_options"),
                                                        hr(),
                                                        uiOutput("donor"),
                                                        hr(),
                                                        uiOutput("recipient"),
                                                        hr(),
                                                        uiOutput("post_fmt"),
                                                        actionButton("split_into_experiment", "Set/Update FMT Details"),
                                                        hr(),
                                                        span(strong(textOutput('second_check')), style = "color:green"),
                                                        hr(),
                                                        span(strong(textOutput('num_donor_samples')), style = "color:blue"),
                                                        span(strong(textOutput('num_recipient_samples')), style = "color:red"),
                                                        span(strong(textOutput('num_post_fmt_samples')), style = "color:green"),
														bsPopover('split_into_experiment', "Select or Update FMT Details", "Select or Update the Donor, Recipient, and Post-FMT metadata categories from the selected column of the mapping file.", placement = 'top')
                                                        )
                                              )
                             ),
                      column(6,
                             conditionalPanel(condition = "output.second_check == 'Experiment-specific table created...'",
                                              wellPanel(h3("3) Normalize and Filter Samples"),
                                                        numericInput("min_OTU_fraction", label = h5("Minimum Relative Abundance Filter"), min = 0.0001, max = 0.1, value = 0.001),
														actionButton('update_min_fraction_filter', 'Apply/Update Filter'),
                                                        bsPopover('min_OTU_fraction', "Relative Abundance Filter", "The minimum fraction that an OTU must represent in a sample (in the sample it is most abundant), in order to be kept. (i.e. an OTU must represent at least this fraction of the microbiota of any one sample in order to be kept)"),
                                                        hr(),
                                                        span(strong(textOutput('third_check')), style = "color:green"),
                                                        hr(),
                                                        h5("Resultant OTUs:"),
                                                        textOutput("n_otus_preprocess"))
                             ),
                             conditionalPanel(condition = "output.third_check == 'Experiment-specific table has been normalized and filtered...'",
                                              wellPanel(h3("4) Remove fleeting OTUs"),
                                                        sliderInput("min_fraction", label = h5("Minimum fraction of samples with a non-zero abundance"), min = 0, max = 1.0, value = 2/3),
                                                        bsPopover("min_fraction", 'Minimum Fraction', "The minimum fraction of samples within a condition that must have a non-zero abundance of an OTU in order for it to be retained as reliably present in that condition."),
                                                        h5("Resultant OTUs:"),
                                                        textOutput("num_otus_after_nonzero_filter")
                                              )),
                             conditionalPanel(condition = "output.num_otus_after_nonzero_filter != null",
                                              wellPanel(h3("5) Select other settings"),
                                                        selectInput("comparison_test", label = h5("Select metric to compare abundances"), choices = list("Median" = "Median", "Mean" = "Mean"), selected = "Median"),
                                                        hr(),
                                                        h5("Exclude post-transplant unique & shared OTUs?"),
                                                        checkboxInput("remove_OTUs_test_specific", "Exclude", value = FALSE),
														bsPopover('remove_OTUs_test_specific', "Exclude?", "Exclude OTUs that are only in the post-transplant samples (i.e. neither in the donor nor the pre-transplant recipient samples), or shared in all conditions (donor, recipient and post-transplant recipient)"),
				 									   hr(),
				 									   checkboxInput('add_tax_UI', 'Add Taxonomy?', value = TRUE),
				 						               uiOutput('Tax_add'),
				 									   bsPopover('id_tax_file', "OTU ID-to-Taxonomy Map", "A .txt file that contains the taxonomy associated with each OTU ID. This can be from the greengenes reference, for example, or from the output of the assign_taxonomy.py function in QIIME."),
				 									   hr(),
				 						               span(strong(textOutput('tax_loaded')), style = "color:green"),
				 									   hr(),
				 									   uiOutput('scale_biomass')
                                                       )
                             ))
             )

    ),
    tabPanel("View Current Inputs & Settings",
             h2("Current Inputs & Settomgs"),
             helpText("Current inputs and settings. In order to change them, navigate to the 'Select Inputs & Settings' page."),
             tableOutput('settings'),
             hr(),
             p(class = "text-center", downloadButton('x0', 'Download This Table')) ),

    "Taxonomy-Independent Output",

    tabPanel("Transplant Metrics by OTU",
             h2("Transplant Metrics by OTU"),
             helpText("Core set of metrics to help gauge: 1) how effective the donor microbiota is in taking hold in the recipient, 2) how receptive the recipient microbiota is to a transplanted microbiota, and 3) the numbers of unique and specific OTUs for the donor and recipient microbiotas."),
             fluidRow(column(3, h4("Donor"),
                             textOutput("donor_id")),
                      column(3, h4("Recipient"),
                             textOutput("recipient_id")),
                      column(3, h4("Post-FMT"),
                             textOutput("post_fmt_id")),
                      column(3, h4("Unique & Shared Removed?"),
                             textOutput("remove_OTUs"))
                      ),
             hr(),
             fluidRow(
               column(6,
                      wellPanel(style = 'overflow:hidden;background-color:white',
					  h4("Numbers of OTUs in each sample set"),
					  br(),
                                uiOutput("N_Donor"),
								bsPopover('N_Donor', title = HTML(paste("N", tags$sub("Donor"), sep = "")), "The number of OTUs that are reliably and exclusively in the donor.", placement = 'top'),
								br(),
								uiOutput("N_Recipient"),
								bsPopover('N_Recipient', title = HTML(paste("N", tags$sub("Recipient"), sep = "")), "The number of OTUs that are reliably and exclusively in the recipient.", placement = 'top'),
								br(),
								uiOutput("N_P_Total"),
								bsPopover('N_P_Total', title = HTML(paste("N", tags$sub("Post-FMT"), sep = "")), "The number of OTUs that are reliably in the post-transplant samples (minus number that are unique to the post-transplant samples and the number that are shared throughout, if this is selected for).", placement = 'top'),
								br(),
								uiOutput("N_P_Unique"),
								bsPopover('N_P_Unique', title = HTML(paste("N", tags$sub("Post-FMT (Unique)"), sep = "")), "Number of OTUs that are reliably and exclusively in the post-transplant samples. This number should theoretically be low or zero, but can serve as a useful pseudo-negative control. These OTUS were in NEITHER the Donor NOR the Recipient microbiotas.", placement = 'top'),
								br(),
								uiOutput("N_P_Shared"),
								bsPopover('N_P_Shared', title = HTML(paste("N", tags$sub("Post-FMT (Shared)"), sep = "")), "Number of OTUs that are reliably present in both donor AND recipient samples throughout.", placement = 'top'),
								br(),
		   					 uiOutput("N_P_Donor"),
		   					 bsPopover("N_P_Donor", title = HTML(paste("N", tags$sub("Donor | Post-FMT"), sep = "")), content = "The number of OTUs in the post-transplant samples that came from the donor.", placement = 'top'),
							 br(),
					 uiOutput("N_P_Recipient"),
					 bsPopover("N_P_Recipient", title = HTML(paste("N", tags$sub("Recipient | Post-FMT"), sep = "")), content = "The number of OTUs in the post-transplant samples that came from the recipient.", placement = 'top')
								)),
				column(6, 
					 wellPanel(style = 'overflow:hidden;background-color:white',
					 h4("FMT Metrics"),
					 br(),
					 uiOutput("D_Engraft"),
					 bsPopover('D_Engraft', title = HTML(paste("D", tags$sub("Engraft"), sep = "")), content = HTML(paste("The proportion of the donor microbiota that engrafts into the recipient.", tags$img(src= 'D-engraft.png'), sep = "")), placement = 'top'),
					 br(),
					 uiOutput("R_Persist"),
					 bsPopover('R_Persist', title = HTML(paste("R", tags$sub("Persist"), sep = "")), content = HTML(paste("The proportion of the recipient microbiota that persists after the transplant.", tags$img(src = 'R-persist.png'), sep = "")), placement = 'top'),					 
					 br(),
					 uiOutput("P_Donor"),
					 bsPopover("P_Donor", title = HTML(paste("P", tags$sub("Donor"), sep = "")), content = HTML(paste("The proportion of OTUs in the post-transplant samples that is derived from the donor.", tags$img(src = 'P-donor.png'), sep = "")), placement = 'top'),
					 br(),
					 uiOutput("P_Recipient"),
					 bsPopover("P_Recipient", title = HTML(paste("P", tags$sub("Recipient"), sep = "")), content = HTML(paste("The proportion of OTUs in the post-transplant samples that is derived from the recipient.", tags$img(src = 'P-recipient.png'), sep = "")), placement = 'top'),
					 br(),
					 uiOutput("P_Unique"),
					 bsPopover("P_Unique", title = HTML(paste("P", tags$sub("Unique"), sep = "")), content = HTML(paste("The proportion of OTUs in the post-transplant samples that are neither in the donor NOR in the recipient.", tags$img(src = 'P-unique.png'), sep = "")), placement = 'top'),
					 br(),
					 uiOutput("P_Shared"),
					 bsPopover("P_Shared", title = HTML(paste("P", tags$sub("Shared"), sep = "")), content = HTML(paste("The proportion of OTUs in the post-transplant samples that are shared in both the donor AND the recipient.", tags$img(src = 'P-shared.png'), sep = "")), placement = 'top')))
             ),
             hr(),
             h4("Transplant Vizualization"),
             plotOutput("metric_visualization")
             ,
             p(class = "text-center", downloadButton('metric_vis_download', "Download Plot as .png")),
             hr(),
             tableOutput('metrics'),
             p(class = "text-center", downloadButton('xmetrics', 'Download As A Table'))

    ),

    tabPanel("Interactive Ordination Plot (By OTU)",
             h2("Interactive Ordination Plot (By OTU)"),

             fluidRow(column(6,
                             wellPanel(selectInput('distance_method', label = h4("Distance Method"), choices = list("Euclidean" = "euclidean", "Manhattan" = "manhattan", "Canberra" = "canberra", "Binary" = "binary", "Minkowski" = "minkowski"), selected = "euclidean"),
                                       uiOutput('plot_x'),
                                       uiOutput('plot_y'),
                                       uiOutput('plot_color_by'))
             ),
             column(6,
                    wellPanel(uiOutput('plot_facet_row'),
                              uiOutput('plot_facet_col'),
                              checkboxInput('jitter', 'Jitter'),
                              checkboxInput('smooth', 'Smooth'),
                              sliderInput("point_size", "Size:",
                                          min = 2, max = 10, value = 6, step = 1))

             ),
             hr(),
             fluidRow(column(12,
                             plotOutput("plot1"))),
             hr(),
			 fluidRow(column(12,
				 			 plotOutput('spree_plot'))),

			 hr(),
             fluidRow(column(2,
                             h4("PC1 Percent Explained"),
                             textOutput("pc1")),
                      column(2,
                             h4("PC2 Percent Explained"),
                             textOutput("pc2")),
                      column(2,
                             h4("PC3 Percent Explained"),
                             textOutput("pc3")),
                      column(2,
                             h4("PC4 Percent Explained"),
                             textOutput("pc4")),
                      column(2,
                             h4("PC5 Percent Explained"),
                             textOutput("pc5"))),
             hr(),
             p(class = "text-center", downloadButton('plot_download', "Download Plot as .png")),
             p(class = "text-center", downloadButton('pc_table_download', "Download Table with Data"))
             )
    ),
	
    tabPanel("OTUs Unique to Each Condition",
             h2("OTUs Unique to Each Condition"),
             p("A data table containing only OTUs that pass the following filtering criteria. 1) The OTU must be present in at least the fraction of samples of a condition as designated in the settings. 2) If the OTU is present in the donor and not in the recipient, it is kept, and designated as 'specific' to the donor. If the OTU is present in the recipient and not in the donor, it is kept and designated as 'specific' to the recipient. If the OTU is present only in the post-FMT samples, but in neither the donor nor the recipient, it is kept, and designated as specific to the post-FMT samples. Theoretically there would be no OTUs specific for the post-FMT samples, and thus this serves as some sort of negative control."),
			 h4("Donor-Specific OTUs"),
			 actionButton("showhide_donor", "Show/Hide"),
			 dataTableOutput('donor_unique'),
			 hr(),
			 h4("Recipient-Specific OTUs"),
			 actionButton("showhide_recipient", "Show/Hide"),
			 dataTableOutput('recipient_unique'),
			 hr(),
			 h4('OTUs Shared Throughout'),
			 actionButton("showhide_shared", "Show/Hide"),
			 dataTableOutput('shared_throughout'),
			 hr(),
			 h4('Post-FMT Unique OTUs'),
			 actionButton("showhide_post_fmt", "Show/Hide"),
			 dataTableOutput('post_fmt_unique'),
             hr()),
	

    tabPanel("OTU Abundances",
             h2("OTU Abundances"),
             p("A data table containing only the OTUs that are unique to each condition, and the mean or median abundance of that OTU in the samples of that condition. This is meant to give a sense of the relative abundance of an OTU beyond its simple presence/absence that is captured by their being unique to a particular condition. Further, the pairwise differences of these abundances are also tabulated between the donor, recipient and post-FMT samples. This helps identify whether an OTU is abundant at comperable levels between the donor and post-FMT samples."),
             hr(),
             fluidRow(column(6, h4("Chosen Abundance Comparison Test Metric"),
                             textOutput("test_metric")),
                      column(6,
                             sliderInput("min_diff", label = h4("Minimum difference to display"), min = 0, max = 0.3, value = 0),
                             helpText("Filters the output abundance tables to exclude OTUs that do not have a minimum difference of abundances between sample conditions (NOTE: minimum or equal to)"))
             ),
             hr(),
             dataTableOutput("filtered_relative_abundance_of_unique_OTUs_plus_differences"),
             hr(),
             p(class = "text-center", downloadButton('x5', 'Download This Table'))),

   

    tabPanel("Raw OTU Table + Metadata",
             h2("Raw OTU Table + Metadata"),
             helpText("A data table containing the raw OTU information plus the added mapping file metadata for all samples fulfilling the comparison criteria inputted."),
             dataTableOutput("experiment_specific_full_table"),
             hr(),
             p(class = "text-center", downloadButton('x1', 'Download This Table')) ),

    tabPanel("Fraction of Samples with Non-Zero Abundances for Each OTU",
             h2("Fraction of Samples with Non-Zero Abundances for Each OTU"),
             helpText("A list of OTUs, the fraction of samples in each condition that have non-zero abundances for that OTU. The fraction represents the fraction of samples within the corresponding 'specificity' condition. This list is filtered according to the minimum fraction setting provided (default = 2/3)."),
             dataTableOutput("fraction_samples_with_nonzero_values_for_OTUs"),
             hr(),
             p(class = "text-center", downloadButton('x2', 'Download This Table'))),

    "Taxonomy-Dependent Output",

    tabPanel("Add Taxonomic Information",
             p('Taxonomic information is added to the individually created tables for the donor, recipient, and post-transplant recipient, containing only stably present OTUs. Then, OTUs that represent the same taxonomic group are collapsed into one, with the resulting abundance information representing the average abundance of all OTUs that correspond to that taxa. These tables now contain ')
             ),

    tabPanel("Transplant Metrics by Taxa",
             h2("Transplant Metrics by Taxa"),
             helpText("Core set of metrics to help gauge: 1) how effective the donor microbiota is in taking hold in the recipient, 2) how receptive the recipient microbiota is to a transplanted microbiota, and 3) the numbers of unique and specific taxa for the donor and recipient microbiotas."),
             fluidRow(column(3, h4("Donor"),
                             textOutput("donor_id_taxa")),
                      column(3, h4("Recipient"),
                             textOutput("recipient_id_taxa")),
                      column(3, h4("Post-FMT"),
                             textOutput("post_fmt_id_taxa")),
                      column(3, h4("Unique & Shared Removed?"),
                             textOutput("remove_OTUs_taxa"))
             ),
             hr(),
             fluidRow(
               column(6,
                      wellPanel(style = 'overflow:hidden;background-color:white',
					  h4("Numbers of Taxa in each sample set"),
					  br(),
                                uiOutput("N_Donor_taxa"),
								bsPopover('N_Donor_taxa', title = HTML(paste("N", tags$sub("Donor"), sep = "")), "The number of Taxa that are reliably and exclusively in the donor.", placement = 'top'),
								br(),
								uiOutput("N_Recipient_taxa"),
								bsPopover('N_Recipient_taxa', title = HTML(paste("N", tags$sub("Recipient"), sep = "")), "The number of Taxa that are reliably and exclusively in the recipient.", placement = 'top'),
								br(),
								uiOutput("N_P_Total_taxa"),
								bsPopover('N_P_Total_taxa', title = HTML(paste("N", tags$sub("Post-FMT"), sep = "")), "The number of Taxa that are reliably in the post-transplant samples (minus number that are unique to the post-transplant samples and the number that are shared throughout, if this is selected for).", placement = 'top'),
								br(),
								uiOutput("N_P_Unique_taxa"),
								bsPopover('N_P_Unique_taxa', title = HTML(paste("N", tags$sub("Post-FMT (Unique)"), sep = "")), "Number of Taxa that are reliably and exclusively in the post-transplant samples. This number should theoretically be low or zero, but can serve as a useful pseudo-negative control. These Taxa were in NEITHER the Donor NOR the Recipient microbiotas.", placement = 'top'),
								br(),
								uiOutput("N_P_Shared_taxa"),
								bsPopover('N_P_Shared_taxa', title = HTML(paste("N", tags$sub("Post-FMT (Shared)"), sep = "")), "Number of Taxa that are reliably present in both donor AND recipient samples throughout.", placement = 'top'),
								br(),
		   					 uiOutput("N_P_Donor_taxa"),
		   					 bsPopover("N_P_Donor_taxa", title = HTML(paste("N", tags$sub("Donor | Post-FMT"), sep = "")), content = "The number of Taxa in the post-transplant samples that came from the donor.", placement = 'top'),
							 br(),
					 uiOutput("N_P_Recipient_taxa"),
					 bsPopover("N_P_Recipient_taxa", title = HTML(paste("N", tags$sub("Recipient | Post-FMT"), sep = "")), content = "The number of Taxa in the post-transplant samples that came from the recipient.", placement = 'top')
								)),
				column(6, 
					 wellPanel(style = 'overflow:hidden;background-color:white',
					 h4("FMT Metrics"),
					 br(),
					 uiOutput("D_Engraft_taxa"),
					 bsPopover('D_Engraft_taxa', title = HTML(paste("D", tags$sub("Engraft"), sep = "")), content = HTML(paste("The proportion of the donor microbiota that engrafts into the recipient.", tags$img(src= 'D-engraft.png'), sep = "")), placement = 'top'),
					 br(),
					 uiOutput("R_Persist_taxa"),
					 bsPopover('R_Persist_taxa', title = HTML(paste("R", tags$sub("Persist"), sep = "")), content = HTML(paste("The proportion of the recipient microbiota that persists after the transplant.", tags$img(src = 'R-persist.png'), sep = "")), placement = 'top'),					 
					 br(),
					 uiOutput("P_Donor_taxa"),
					 bsPopover("P_Donor_taxa", title = HTML(paste("P", tags$sub("Donor"), sep = "")), content = HTML(paste("The proportion of Taxa in the post-transplant samples that is derived from the donor.", tags$img(src = 'P-donor.png'), sep = "")), placement = 'top'),
					 br(),
					 uiOutput("P_Recipient_taxa"),
					 bsPopover("P_Recipient_taxa", title = HTML(paste("P", tags$sub("Recipient"), sep = "")), content = HTML(paste("The proportion of Taxa in the post-transplant samples that is derived from the recipient.", tags$img(src = 'P-recipient.png'), sep = "")), placement = 'top'),
					 br(),
					 uiOutput("P_Unique_taxa"),
					 bsPopover("P_Unique_taxa", title = HTML(paste("P", tags$sub("Unique"), sep = "")), content = HTML(paste("The proportion of Taxa in the post-transplant samples that are neither in the donor NOR in the recipient.", tags$img(src = 'P-unique.png'), sep = "")), placement = 'top'),
					 br(),
					 uiOutput("P_Shared_taxa"),
					 bsPopover("P_Shared_taxa", title = HTML(paste("P", tags$sub("Shared"), sep = "")), content = HTML(paste("The proportion of Taxa in the post-transplant samples that are shared in both the donor AND the recipient.", tags$img(src = 'P-shared.png'), sep = "")), placement = 'top')))
             ),
             hr(),
             h4("Visualization of Transplant"),
			 plotOutput("metric_visualization_taxa"),
			 hr(),
             p(class = "text-center", downloadButton('metric_vis_download_taxa', "Download Plot as .png")),
             hr(),
             tableOutput('metrics_taxa'),
             p(class = "text-center", downloadButton('xmetrics_taxa', 'Download As A Table'))
    ),
	
    tabPanel("Taxa Unique to Each Condition",
             h2("Taxa Unique to Each Condition"),
             p("A data table containing only Taxa that pass the following filtering criteria. 1) The Taxa must be present in at least the fraction of samples of a condition as designated in the settings. 2) If the Taxa is present in the donor and not in the recipient, it is kept, and designated as 'specific' to the donor. If the Taxa is present in the recipient and not in the donor, it is kept and designated as 'specific' to the recipient. If the Taxa is present only in the post-FMT samples, but in neither the donor nor the recipient, it is kept, and designated as specific to the post-FMT samples. Theoretically there would be no Taxa specific for the post-FMT samples, and thus this serves as some sort of negative control."),
			 hr(),
			 h4("Donor-Specific Taxa"),
			 actionButton("showhide_donor_taxa", "Show/Hide"),
			 dataTableOutput('donor_unique_taxa'),
			 hr(),
			 h4("Recipient-Specific Taxa"),
			 actionButton("showhide_recipient_taxa", "Show/Hide"),
			 dataTableOutput('recipient_unique_taxa'),
			 hr(),
			 h4('Taxa Shared Throughout'),
			 actionButton("showhide_shared_taxa", "Show/Hide"),
			 dataTableOutput('shared_throughout_taxa'),
			 hr(),
			 h4('Post-FMT Unique Taxa'),
			 actionButton("showhide_unique_taxa", "Show/Hide"),
			 dataTableOutput('post_fmt_unique_taxa'),
             hr()
             #p(class = "text-center", downloadButton('x3_taxa', 'Download This Table'))
			 ),
	

    tabPanel("OTU Abundances + Taxonomy",
             h3("OTU Abundances + Taxonomy"),
             helpText("The same OTU Abundances table as above, but with the taxonomic information added."),
             dataTableOutput("taxonomy_added_final"),
             helpText("NOTE: Since adding the taxonomic information takes a while, you will have to click on the 'Add/Refresh Taxonomy to Tables' button in order to refresh or add the taxonomy to the tables if you have made any changes to the taxonomy-independent tables."),
             hr(),
             p(class = "text-center", downloadButton('x7', 'Download This Table')) ),

    tabPanel("Collapsed By Taxonomy",
             h3("Collapsed By Taxonomy"),
             helpText("A table that collapses the 'OTU Abundances + Taxonomy' table on the basis of Taxonomy. This is useful since multiple OTUs may map to the same taxonomic group. Since multiple numbers are combined in this operation, the values in the table represent a simple average of their values"),
             dataTableOutput("collapsed_taxonomy_table"),
             helpText("NOTE: Since adding the taxonomic information takes a while, you will have to click on the 'Add/Refresh Taxonomy to Tables' button in order to refresh or add the taxonomy to the tables if you have made any changes to the taxonomy-independent tables."),
             hr(),
             p(class = "text-center", downloadButton('x9', 'Download This Table'))),

    tabPanel("Distinct Taxa Summary Table",
             h3("Distinct Taxa Summary Table"),
             helpText("A table that summarizes how many distinct taxa are in each condition. Since multiple OTUs may map to the same taxonomic group, this helps give a sense of how many taxonomically distinct groups actually are present in each sample."),
             dataTableOutput("distinct_taxa_summary_table"),
             helpText("NOTE: Since adding the taxonomic information takes a while, you will have to click on the 'Add/Refresh Taxonomy to Tables' button in order to refresh or add the taxonomy to the tables if you have made any changes to the taxonomy-independent tables."),
             hr(),
             p(class = "text-center", downloadButton('x8', 'Download This Table')) )

  )



))